Looks up for `kid` and `alg` also in unprotected header

lmammino opened the pull request:
Fixes #1 

- Parses `kid` and `alg` from unprotected header. 
- Uses SHA256 hash to calculate kid from certificates
- Removes `header_unprotected` (not used anymore) and makes `header_protected_raw` and `payload_raw` private in `Cwt` (it felt like a leaking abstraction)
- Makes sure all the tests originally skipped because of #1 now run correctly

~~Note: It is better to merge #23 first and then rebase this branch, so we can easily test against all the previously failing tests.~~ Done

CC: @rez23 who was already looking at this!

lmammino committed:
Parses kid and alg from unprotected header. Uses SHA256 hash to calculate kid from certificates

lmammino committed:
console.log like there is no tomorrow

lu-zero commented on src/cwt.rs:
```diff
@@ -167,28 +151,32 @@ impl TryFrom<&[u8]> for Cwt {
         let payload_raw = (parts[2].as_bytes().ok_or(PayloadNotBinary)?).clone();
         let signature = (parts[3].as_bytes().ok_or(SignatureNotBinary)?).clone();
 
+        let unprotected_header_iter = parts[1].as_map().unwrap().iter();
```
when does the unwrap() fail? you could do as you do in line 163 as well, I think.

lmammino replied:
Good catch, let me review these nasty `unwrap()`s

lmammino replied:
Addressed!

lu-zero reviewed (changes requested):
You may squash the two commits in one while at it.

lmammino committed:
Removed unwrap and header_unprotected from `Cwt` (not used)

lmammino committed:
Better parsing of headers

lmammino commented on src/dgc_cert.rs:
```diff
@@ -13,19 +13,52 @@ pub struct DgcCertName {
     pub fnt: String,
 }
 
+fn t_empty_if_null<'de, D>(deserializer: D) -> Result<Vec<Test>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let opt = Option::deserialize(deserializer)?;
+    Ok(opt.unwrap_or_default())
+}
+
+fn v_empty_if_null<'de, D>(deserializer: D) -> Result<Vec<Vaccination>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let opt = Option::deserialize(deserializer)?;
+    Ok(opt.unwrap_or_default())
+}
+
+fn r_empty_if_null<'de, D>(deserializer: D) -> Result<Vec<Recovery>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let opt = Option::deserialize(deserializer)?;
+    Ok(opt.unwrap_or_default())
+}
+
```
Note that I had to do this because there were 2 tests (originally skipped because of the "kid in unprotected header" problem) that have the following content encoded:

```json
 {
    "r": null,
    "t": null,
    "v":  [
       // ...
    ],
}
```

Instead of having only one value set between `r`, `t`, or `v` they set two of them to `null` and the other to an appropriate array.

Solution based on https://github.com/serde-rs/serde/issues/1098

lmammino replied:
I have generalised this to one generic `empty_if_null` function...

dodomorandi replied:
FYI: https://docs.rs/serde_with/1.11.0/serde_with/
This can be pretty useful...

lmammino replied:
Nice one. Is there any utility there that would do what we are doing here? I couldn't find one at first glance...

dodomorandi replied:
In this case I am not sure. However, this a lifesaver to avoid `skip_serialization_if(Option::is_none)` everywhere. For some of my project I discovered this a bit too late... :sweat:

lmammino commented:
@lu-zero, this is now rebases with `main`. I can do a "Squash and merge" once this is ready to go rather than squashing everything upfront. Would that work for you?

lmammino commented:
Note that with this PR we solve most of the broken test. We go from:

```plain
test result: ok. 528 passed; 0 failed; 52 ignored; 0 measured; 0 filtered out; finished in 0.17s
```


To:

```plain
test result: ok. 558 passed; 0 failed; 22 ignored; 0 measured; 0 filtered out; finished in 0.17s
```

So **30** _new_ tests are now passing! 🎉

lmammino committed:
Generalised x_empty_if_null into one generic function

dodomorandi reviewed (commented):
I am going to propose a commit (it would be a little bit a mess using diff changes), feel free to reject it if you find it out of scope. I noticed that some parts have been refactored but they could be improved a little bit.

dodomorandi committed:
Directly use owned data instead of cloning

Unfortunately `ciborium` misses some ergonomic features in order to keep
things easy. I added an _extension trait_ to overcome this problem and I
used it to keep code a little bit cleaner.

dodomorandi commented:
[This](https://github.com/rust-italia/dgc/pull/25/commits/b4e22f328dad1c2f10c90c538041eba38e98f3fc) is the change I was talking about. Let me know what do you think about it.

guerinoni reviewed (approved):
LGTM :)

lmammino commented:
I love the suggested changes @dodomorandi! Thanks. 

I actually contributed to ciborium adding the existing `as_x` functions to `Value`. It would be nice to submit a PR with these new features. You should do it 😉

I'll try to fix the conflict, rebase and get this merged today 🤞

dodomorandi commented:
> I'll try to fix the conflict, rebase and get this merged today :crossed_fingers:

No worries, I'll do it because I caused the conflict :grin:. I was just waiting for your feedback before performing a forced push.

> It would be nice to submit a PR with these new features. You should do it :wink:

I totally agree!!
